File: matlab/EnableGPU.m
function EnableGPU(idx)
    try
        if gpuDeviceCount("available") >= 1
            if nargin < 1
                idx = 1;
            end
            g = gpuDevice(idx);
            fprintf('Using GPU %d: %s\n', g.Index, g.Name);
        else
            fprintf('No CUDA GPU detected. Using CPU.\n');
        end
    catch ME
        warning('EnableGPU:InitFailed', ...
            'GPU init failed: %s. Continuing on CPU.', ME.message);
    end
end

File: resnet18_ensembles/scripts/enable_gpu.py
import torch

def enable_gpu(idx: int = 0):
    """Select a GPU (like EnableGPU.m) and print info."""
    try:
        if torch.cuda.is_available():
            device = torch.device(f"cuda:{idx}")
            gpu_name = torch.cuda.get_device_name(idx)
            free, total = torch.cuda.mem_get_info()
            print(f"[GPU] Using {gpu_name} ({free/1e9:.1f} GB free / {total/1e9:.1f} GB total)")
        else:
            device = torch.device("cpu")
            print("[GPU] No CUDA GPU detected. Using CPU.")
    except Exception as e:
        print(f"[WARN] GPU init failed: {e}. Falling back to CPU.")
        device = torch.device("cpu")
    return device